go/mysql/conn.go

```diff
@@ -548,17 +551,29 @@ func (c *Conn) startEphemeralPacket(length int) []byte {
 	return c.currentEphemeralPacket
 }
 
+func (c *Conn) endEphemeralPacket() {
+	c.currentEphemeralPolicy = ephemeralUnused
+}
+
 // writeEphemeralPacket writes the packet that was allocated by
 // startEphemeralPacket. If 'direct' is set, we write to the
 // underlying connection directly, by-passing the write buffer.
 func (c *Conn) writeEphemeralPacket(direct bool) error {
-	defer func() {
-		c.currentEphemeralPolicy = ephemeralUnused
-	}()
-
+	// TODO(sougou): this call is not necessary. Remove after
+	// tests are fixed
+	defer c.endEphemeralPacket()
 	var w io.Writer = c.writer
+	if w == nil {
+		panic("w (io.Writer) is nil")
+	}
 	if direct {
 		w = c.conn
+		if w == nil {
+			panic("w (c.conn) is nil")
+		}
+	}
+	if c.currentEphemeralBuffer == nil {
```

Comment:
this check is at the wrong place, ephemeralWriteGlobalBuffer policy doesn't use this buffer.